### DISCUSSION TITLE
Is Zwift speed and incline control possible via Raspberry PI?

### DISCUSSION THREAD

**Discussion by magicdude4eva** [cagnulein/qdomyos-zwift#2939] [Q&A] [answered]
I have successfully managed to install QZ on the Raspberry PI via https://github.com/cagnulein/qdomyos-zwift/blob/master/docs/10_Installation.md 

In the past I used the QZ app on IOS and my Nordictrack S25 did work with the Zwift control by configuring the Zwift credentials as via https://github.com/cagnulein/qdomyos-zwift/wiki/Can-Zwift-control-speed-and-or-inclination-of-my-treadmill%3F

Now my question is: Does this work on Raspberry PI? and what settings do I need to make in qDomyos-Zwift.conf (at the moment I have set zwift_password / zwift_username)

I am not sure if it suffices to change the controller name to "KICKR RUN" or if I need to do something else:
```
root@KICKR-RUN:~/.config/Roberto Viola # bluetoothctl
Agent registered
[CHG] Controller D8:3A:DD:B0:17:F6 Pairable: yes
[bluetooth]# show
Controller D8:3A:DD:B0:17:F6 (public)
	Name: KICKR RUN
	Alias: KICKR RUN
	Class: 0x00000000
```

**Comment by cagnulein** [comment]
I didn't port the zwift api to raspberry yet but it's doable, I have only to add the right libraries version to the binary 

For the kickr run trick it should work, did you already try? If it doesn't work (this is the best way instead of the zwift api), could you please show me how you see the raspberry from nrfconnect on the other device?

**Reply by magicdude4eva** [reply]
I have used the NFR Connect iOS App and I am unsure what to look for. The Raspberry PI shows up as a "DomyosBridge" device in the app:
![IMG_0523](https://github.com/user-attachments/assets/38bad59c-6a79-484d-bdc4-3cc779f31f52)

I honestly do not know what to look for:
![IMG_0524](https://github.com/user-attachments/assets/a0f1b6a9-291f-4faf-bf54-54706889dbe0)

I have attached some exported logs, but do not see anything obvious:
[DomyosBridge 3.txt](https://github.com/user-attachments/files/18254836/DomyosBridge.3.txt)
[DomyosBridge 2.txt](https://github.com/user-attachments/files/18254837/DomyosBridge.2.txt)
[DomyosBridge.txt](https://github.com/user-attachments/files/18254838/DomyosBridge.txt)

Perhaps an unrelated question: When running QZ as a service, I see a lot of those messages (treadmill is currently off) - is there a way to surpress this or perhaps still a config issue:
```
root@KICKR-RUN:~ # journalctl -u qz -f
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: Thu Dec 26 19:52:39 2024 1735239159449 Debug:   BTLE_DBUS::connect() failed "org.bluez.Error.Failed" "Operation already in progress"
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: qt.bluetooth.bluez: BTLE_DBUS::connect() failed "org.bluez.Error.Failed" "Operation already in progress"
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: Thu Dec 26 19:52:39 2024 1735239159449 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString&) "proformtreadmill::errorUnknownErrorUnknown Error"
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: "proformtreadmill::errorUnknownErrorUnknown Error"
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: Thu Dec 26 19:52:39 2024 1735239159449 Debug: devices/bluetooth.cpp void bluetooth::debug(const QString&) "Cannot connect to remote device."
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: "Cannot connect to remote device."
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: Thu Dec 26 19:52:39 2024 1735239159450 Debug:   QLowEnergyControllerPrivate setting state to QLowEnergyController::UnconnectedState
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: qt.bluetooth: QLowEnergyControllerPrivate setting state to QLowEnergyController::UnconnectedState
Dec 26 19:52:39 KICKR-RUN qdomyos-zwift[7563]: Thu Dec 26 19:52:39 2024 1735239159450 Debug: devices/proformtreadmill/proformtreadmill.cpp void proformtreadmill::controllerStateChanged(QLowEnergyController::ControllerState) "controllerSta
```

I would appreciate some hints on what to look for - attached also the current version of the config file:
[qDomyos-Zwift.conf.txt](https://github.com/user-attachments/files/18254892/qDomyos-Zwift.conf.txt)

**Reply by cagnulein** [reply]
I started a PR for this
The problem is the Bluetooth name from the raspberry 

Let me know 

https://github.com/cagnulein/qdomyos-zwift/pull/2941

**Reply by magicdude4eva** [reply]
Perfect, this worked and I see the KICKR RUN now:
![IMG_0532](https://github.com/user-attachments/assets/3eef3208-ce8b-4b73-a8c0-00d45a766095)

pairing works and on the main-screen it displays the incline if i adjust it manually on the treadmill (Nordictrack S25):
![IMG_0533](https://github.com/user-attachments/assets/aa257b8f-e049-4ee3-9e97-1f39f6a323e9)

I enabled "Incline Control" in the hardware settings. I did not do any extensive tests, I just started a regular hill-run and did not notice that the incline was adjusted via Zwift. Do I need to adjust any settings in the conf file or is there something else I need to consider?

**Reply by cagnulein** [reply]
you shouldn't enable anything else. could you share a debug log so i can check it?

**Reply by magicdude4eva** [reply]
in [qz-hill-run-kickr.txt.zip](https://github.com/user-attachments/files/18260034/qz-hill-run-kickr.txt.zip) i have done a hill workout and in the video https://share.icloud.com/photos/01axHh5Uh0IvtFaBXbo68qJwA you will see that it does not auto-adjust to 4%.

in [qz-manual-adjustment.txt](https://github.com/user-attachments/files/18260055/qz-manual-adjustment.txt) i did some manual adjustments and the incline reflects on Zwift (video: https://share.icloud.com/photos/0bdsCLEW6uj7Emso0v1QQTdnQ) 

i am unsure if I do this right or if I am missing something?

**Reply by cagnulein** [reply]
@magicdude4eva zwift doesn't send inclination in the workouts (i don't know why, but they developed this in this way). use a free run or a race and you will see the autoinclination

Let me know!

**Comment by magicdude4eva** [comment]
Thanks, I did a group run this morning and saw that Zwift displayed the incline, but the treadmill did not adjust. I will redo this later on with debug logging on and paste the logs.

**Reply by cagnulein** [reply]
ok let me know! did you pair, of course, qz as contrallable, in the pair screen, right?

**Reply by magicdude4eva** [reply]
> ok let me know! did you pair, of course, qz as contrallable, in the pair screen, right?

yes, it was paired and i verified in hardware settings that it is enabled. will post an update later on with logs.

**Reply by magicdude4eva** [reply, marked as answer]
good news: I did a free run and the auto incline works perfectly: https://share.icloud.com/photos/06dN8k5pBpuTwAOXgmxHrBChg

I have also included the debug log in case you want to check something [qz-hill-run-kickr-auto.txt.zip](https://github.com/user-attachments/files/18266051/qz-hill-run-kickr-auto.txt.zip)

thanks for all the good work.

**Reply by cagnulein** [reply]
Ok so probably zwift sends inclination only on free run?

**Reply by magicdude4eva** [reply]
> Ok so probably zwift sends inclination only on free run?

i will see if i can understand when this happens, but does look like yesterday's trainings hill-run did not trigger it.

**Reply by cagnulein** [reply]
Ok so the log that you just attached it's the one where it worked right?

**Comment by magicdude4eva** [comment]
> Ok so the log that you just attached it's the one where it worked right?

Yes. Can we close this issue. I am going to open a separate one regarding the Pi integration.

**Reply by cagnulein** [reply]
ok create a ticket instead of a discussion in case
